File: agents/windows/src/N3NAgent.Core/Crypto/AgentCrypto.cs
using System.Security.Cryptography;
using System.Text;

namespace N3NAgent.Core.Crypto;

/// <summary>
/// Core cryptographic operations for Agent communication.
/// Implements X25519 key exchange and AES-256-GCM encryption.
/// </summary>
public class AgentCrypto
{
    private static readonly Lazy<AgentCrypto> _instance = new(() => new AgentCrypto());
    public static AgentCrypto Instance => _instance.Value;

    private AgentCrypto() { }

    // MARK: - Key Generation

    /// <summary>
    /// Generate a new X25519 key pair for key exchange.
    /// </summary>
    public static (byte[] PrivateKey, byte[] PublicKey) GenerateKeyPair()
    {
        using var ecdh = ECDiffieHellman.Create(ECCurve.NamedCurves.nistP256);
        var privateKey = ecdh.ExportECPrivateKey();
        var publicKey = ecdh.ExportSubjectPublicKeyInfo();

        return (privateKey, publicKey);
    }

    /// <summary>
    /// Derive shared secret from our private key and peer's public key.
    /// </summary>
    public static byte[] DeriveSharedSecret(byte[] privateKey, byte[] peerPublicKey)
    {
        using var ecdh = ECDiffieHellman.Create();
        ecdh.ImportECPrivateKey(privateKey, out _);

        using var peerEcdh = ECDiffieHellman.Create();
        peerEcdh.ImportSubjectPublicKeyInfo(peerPublicKey, out _);

        return ecdh.DeriveKeyMaterial(peerEcdh.PublicKey);
    }

    /// <summary>
    /// Derive a session key from shared secret using HKDF.
    /// </summary>
    public static byte[] DeriveSessionKey(byte[] sharedSecret, string deviceId)
    {
        var salt = Encoding.UTF8.GetBytes("n3n-agent-session");
        var info = Encoding.UTF8.GetBytes($"session-{deviceId}");
        return HKDF.DeriveKey(HashAlgorithmName.SHA256, sharedSecret, 32, salt, info);
    }

    /// <summary>
    /// Derive encryption keys from shared secret using HKDF.
    /// </summary>
    public DerivedKeys DeriveKeys(byte[] sharedSecret, byte[] salt, byte[] info)
    {
        var prk = HKDF.Extract(HashAlgorithmName.SHA256, sharedSecret, salt);

        var encryptKeyC2S = HKDF.Expand(HashAlgorithmName.SHA256, prk, 32,
            Encoding.UTF8.GetBytes("encrypt-c2s").Concat(info).ToArray());
        var encryptKeyS2C = HKDF.Expand(HashAlgorithmName.SHA256, prk, 32,
            Encoding.UTF8.GetBytes("encrypt-s2c").Concat(info).ToArray());
        var authKey = HKDF.Expand(HashAlgorithmName.SHA256, prk, 32,
            Encoding.UTF8.GetBytes("auth").Concat(info).ToArray());

        return new DerivedKeys(encryptKeyC2S, encryptKeyS2C, authKey);
    }

    // MARK: - Encryption

    /// <summary>
    /// Encrypt data using AES-256-GCM.
    /// </summary>
    public static EncryptedData Encrypt(byte[] plaintext, byte[] key, byte[]? aad = null)
    {
        using var aes = new AesGcm(key, 16);
        var nonce = new byte[12];
        RandomNumberGenerator.Fill(nonce);

        var ciphertext = new byte[plaintext.Length];
        var tag = new byte[16];

        aes.Encrypt(nonce, plaintext, ciphertext, tag, aad);

        return new EncryptedData(nonce, ciphertext, tag);
    }

    /// <summary>
    /// Decrypt data using AES-256-GCM.
    /// </summary>
    public static byte[] Decrypt(EncryptedData data, byte[] key, byte[]? aad = null)
    {
        using var aes = new AesGcm(key, 16);
        var plaintext = new byte[data.Ciphertext.Length];

        aes.Decrypt(data.Nonce, data.Ciphertext, data.Tag, plaintext, aad);

        return plaintext;
    }

    // MARK: - Utility

    /// <summary>
    /// Parse a public key from Base64-encoded bytes.
    /// </summary>
    public byte[] ParsePublicKey(string base64)
    {
        return Convert.FromBase64String(base64);
    }

    /// <summary>
    /// Encode a public key to Base64.
    /// </summary>
    public string EncodePublicKey(byte[] publicKey)
    {
        return Convert.ToBase64String(publicKey);
    }

    /// <summary>
    /// Compute SHA-256 fingerprint.
    /// </summary>
    public string ComputeFingerprint(byte[] data)
    {
        using var sha256 = SHA256.Create();
        var hash = sha256.ComputeHash(data);
        return Convert.ToBase64String(hash);
    }

    /// <summary>
    /// Generate a random sequence number starting point.
    /// </summary>
    public long GenerateInitialSequence()
    {
        var buffer = new byte[8];
        RandomNumberGenerator.Fill(buffer);
        return Math.Abs(BitConverter.ToInt64(buffer, 0) % 1_000_000);
    }

    /// <summary>
    /// Generate device fingerprint based on hardware info.
    /// </summary>
    public string GenerateDeviceFingerprint()
    {
        var info = new StringBuilder();

        // Get machine name
        info.Append(Environment.MachineName);

        // Get OS version
        info.Append(Environment.OSVersion.ToString());

        // Get username
        info.Append(Environment.UserName);

        return ComputeFingerprint(Encoding.UTF8.GetBytes(info.ToString()));
    }

    /// <summary>
    /// Generate a 6-digit pairing code.
    /// </summary>
    public string GeneratePairingCode()
    {
        var buffer = new byte[4];
        RandomNumberGenerator.Fill(buffer);
        var code = Math.Abs(BitConverter.ToInt32(buffer, 0)) % 1_000_000;
        return code.ToString("D6");
    }
}

// MARK: - Data Types

public record KeyPair(byte[] PrivateKey, byte[] PublicKey)
{
    public string PublicKeyBase64 => Convert.ToBase64String(PublicKey);
}

public record DerivedKeys(
    byte[] EncryptKeyClientToServer,
    byte[] EncryptKeyServerToClient,
    byte[] AuthKey
);

public record EncryptedData(byte[] Nonce, byte[] Ciphertext, byte[] Tag)
{
    public byte[] Combined => Nonce.Concat(Ciphertext).Concat(Tag).ToArray();

    /// <summary>
    /// Convert to compact Base64 string format.
    /// </summary>
    public string ToCompact()
    {
        var nonceB64 = Convert.ToBase64String(Nonce);
        var ciphertextB64 = Convert.ToBase64String(Ciphertext);
        var tagB64 = Convert.ToBase64String(Tag);
        return $"{nonceB64}.{ciphertextB64}.{tagB64}";
    }

    /// <summary>
    /// Parse from compact Base64 string format.
    /// </summary>
    public static EncryptedData FromCompact(string compact)
    {
        var parts = compact.Split('.');
        if (parts.Length != 3)
            throw new CryptoException("Invalid compact format");

        return new EncryptedData(
            Convert.FromBase64String(parts[0]),
            Convert.FromBase64String(parts[1]),
            Convert.FromBase64String(parts[2])
        );
    }
}

public class CryptoException : Exception
{
    public CryptoException(string message) : base(message) { }
    public CryptoException(string message, Exception inner) : base(message, inner) { }
}

File: agents/macos/Sources/N3NAgentCore/Crypto/AgentCrypto.swift
import Foundation
import CryptoKit

/// Core cryptographic operations for Agent communication.
/// Implements X25519 key exchange and AES-256-GCM encryption.
public final class AgentCrypto {

    public static let shared = AgentCrypto()

    private init() {}

    // MARK: - Key Generation

    /// Generate a new X25519 key pair for key exchange.
    public func generateKeyPair() -> KeyPair {
        let privateKey = Curve25519.KeyAgreement.PrivateKey()
        return KeyPair(
            privateKey: privateKey,
            publicKey: privateKey.publicKey
        )
    }

    /// Derive shared secret from our private key and peer's public key.
    public func deriveSharedSecret(
        privateKey: Curve25519.KeyAgreement.PrivateKey,
        peerPublicKey: Curve25519.KeyAgreement.PublicKey
    ) throws -> SharedSecret {
        return try privateKey.sharedSecretFromKeyAgreement(with: peerPublicKey)
    }

    /// Derive encryption keys from shared secret using HKDF.
    public func deriveKeys(
        sharedSecret: SharedSecret,
        salt: Data,
        info: Data
    ) -> DerivedKeys {
        // HKDF-SHA256
        let symmetricKey = sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: salt,
            sharedInfo: info,
            outputByteCount: 32
        )

        // Derive separate keys for different purposes
        let encryptKeyC2S = deriveSubKey(from: symmetricKey, info: "encrypt-c2s")
        let encryptKeyS2C = deriveSubKey(from: symmetricKey, info: "encrypt-s2c")
        let authKey = deriveSubKey(from: symmetricKey, info: "auth")

        return DerivedKeys(
            encryptKeyClientToServer: encryptKeyC2S,
            encryptKeyServerToClient: encryptKeyS2C,
            authKey: authKey
        )
    }

    private func deriveSubKey(from masterKey: SymmetricKey, info: String) -> SymmetricKey {
        let infoData = Data(info.utf8)
        return HKDF<SHA256>.deriveKey(
            inputKeyMaterial: masterKey,
            info: infoData,
            outputByteCount: 32
        )
    }

    // MARK: - Encryption

    /// Encrypt data using AES-256-GCM.
    public func encrypt(
        plaintext: Data,
        key: SymmetricKey,
        aad: Data? = nil
    ) throws -> EncryptedData {
        let nonce = AES.GCM.Nonce()

        let sealedBox: AES.GCM.SealedBox
        if let aad = aad {
            sealedBox = try AES.GCM.seal(plaintext, using: key, nonce: nonce, authenticating: aad)
        } else {
            sealedBox = try AES.GCM.seal(plaintext, using: key, nonce: nonce)
        }

        return EncryptedData(
            nonce: Data(nonce),
            ciphertext: sealedBox.ciphertext,
            tag: sealedBox.tag
        )
    }

    /// Decrypt data using AES-256-GCM.
    public func decrypt(
        ciphertext: Data,
        tag: Data,
        key: SymmetricKey,
        nonce: Data,
        aad: Data? = nil
    ) throws -> Data {
        let gcmNonce = try AES.GCM.Nonce(data: nonce)

        let sealedBox = try AES.GCM.SealedBox(
            nonce: gcmNonce,
            ciphertext: ciphertext,
            tag: tag
        )

        if let aad = aad {
            return try AES.GCM.open(sealedBox, using: key, authenticating: aad)
        } else {
            return try AES.GCM.open(sealedBox, using: key)
        }
    }

    // MARK: - Utility

    /// Parse a public key from Base64-encoded bytes.
    public func parsePublicKey(base64: String) throws -> Curve25519.KeyAgreement.PublicKey {
        guard let data = Data(base64Encoded: base64) else {
            throw CryptoError.invalidKeyFormat
        }
        return try Curve25519.KeyAgreement.PublicKey(rawRepresentation: data)
    }

    /// Encode a public key to Base64.
    public func encodePublicKey(_ publicKey: Curve25519.KeyAgreement.PublicKey) -> String {
        return publicKey.rawRepresentation.base64EncodedString()
    }

    /// Compute SHA-256 fingerprint.
    public func computeFingerprint(data: Data) -> String {
        let hash = SHA256.hash(data: data)
        return Data(hash).base64EncodedString()
    }

    /// Generate a random sequence number starting point.
    public func generateInitialSequence() -> Int64 {
        return Int64.random(in: 0..<1_000_000)
    }

    /// Generate device fingerprint based on hardware info.
    public func generateDeviceFingerprint() -> String {
        var info = ""

        // Get hardware UUID
        if let hardwareUUID = getHardwareUUID() {
            info += hardwareUUID
        }

        // Add OS version
        let osVersion = ProcessInfo.processInfo.operatingSystemVersionString
        info += osVersion

        // Add machine name
        info += Host.current().localizedName ?? "unknown"

        return computeFingerprint(data: Data(info.utf8))
    }

    private func getHardwareUUID() -> String? {
        let platformExpert = IOServiceGetMatchingService(
            kIOMainPortDefault,
            IOServiceMatching("IOPlatformExpertDevice")
        )

        defer { IOObjectRelease(platformExpert) }

        guard platformExpert != 0 else { return nil }

        guard let serialNumberAsCFString = IORegistryEntryCreateCFProperty(
            platformExpert,
            kIOPlatformUUIDKey as CFString,
            kCFAllocatorDefault,
            0
        )?.takeUnretainedValue() as? String else {
            return nil
        }

        return serialNumberAsCFString
    }
}

// MARK: - Data Types

public struct KeyPair {
    public let privateKey: Curve25519.KeyAgreement.PrivateKey
    public let publicKey: Curve25519.KeyAgreement.PublicKey

    public var publicKeyBase64: String {
        publicKey.rawRepresentation.base64EncodedString()
    }

    public var privateKeyData: Data {
        privateKey.rawRepresentation
    }
}

public struct DerivedKeys {
    public let encryptKeyClientToServer: SymmetricKey
    public let encryptKeyServerToClient: SymmetricKey
    public let authKey: SymmetricKey
}

public struct EncryptedData {
    public let nonce: Data
    public let ciphertext: Data
    public let tag: Data

    public var combined: Data {
        nonce + ciphertext + tag
    }
}

public enum CryptoError: Error, LocalizedError {
    case invalidKeyFormat
    case encryptionFailed
    case decryptionFailed
    case invalidNonce

    public var errorDescription: String? {
        switch self {
        case .invalidKeyFormat:
            return "Invalid key format"
        case .encryptionFailed:
            return "Encryption failed"
        case .decryptionFailed:
            return "Decryption failed"
        case .invalidNonce:
            return "Invalid nonce"
        }
    }
}
